Source/WebKit/WPEPlatform/wpe/WPEBuffer.cpp

```diff
@@ -0,0 +1,233 @@
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WPEBuffer.h"
+
+#include <wtf/glib/GRefPtr.h>
+#include <wtf/glib/WTFGType.h>
+
+/**
+ * WPEBuffer:
+ *
+ */
+struct _WPEBufferPrivate {
+    GRefPtr<WPEDisplay> display;
+    int width;
+    int height;
+};
+
+WEBKIT_DEFINE_ABSTRACT_TYPE(WPEBuffer, wpe_buffer, G_TYPE_OBJECT)
+
+/**
+ * wpe_buffer_error_quark:
+ *
+ * Gets the WPEBuffer Error Quark.
+ *
+ * Returns: a #GQuark.
+ **/
+G_DEFINE_QUARK(wpe-buffer-error-quark, wpe_buffer_error)
+
+enum {
+    PROP_0,
+
+    PROP_DISPLAY,
+    PROP_WIDTH,
+    PROP_HEIGHT,
+
+    N_PROPERTIES
+};
+
+static GParamSpec* sObjProperties[N_PROPERTIES] = { nullptr, };
+
+static void wpeBufferSetProperty(GObject* object, guint propId, const GValue* value, GParamSpec* paramSpec)
+{
+    auto* buffer = WPE_BUFFER(object);
+
+    switch (propId) {
+    case PROP_DISPLAY:
+        buffer->priv->display = WPE_DISPLAY(g_value_get_object(value));
+        break;
+    case PROP_WIDTH:
+        buffer->priv->width = g_value_get_int(value);
+        break;
+    case PROP_HEIGHT:
+        buffer->priv->height = g_value_get_int(value);
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
+    }
+}
+
+static void wpeBufferGetProperty(GObject* object, guint propId, GValue* value, GParamSpec* paramSpec)
+{
+    auto* buffer = WPE_BUFFER(object);
+
+    switch (propId) {
+    case PROP_DISPLAY:
+        g_value_set_object(value, wpe_buffer_get_display(buffer));
+        break;
+    case PROP_WIDTH:
+        g_value_set_int(value, wpe_buffer_get_width(buffer));
+        break;
+    case PROP_HEIGHT:
+        g_value_set_int(value, wpe_buffer_get_height(buffer));
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
+    }
+}
+
+static void wpe_buffer_class_init(WPEBufferClass* bufferClass)
+{
+    GObjectClass* objectClass = G_OBJECT_CLASS(bufferClass);
+    objectClass->set_property = wpeBufferSetProperty;
+    objectClass->get_property = wpeBufferGetProperty;
+
+    /**
+     * WPEBuffer:display:
+     *
+     * The #WPEDisplay of the buffer.
+     */
+    sObjProperties[PROP_DISPLAY] =
+        g_param_spec_object(
+            "display",
+            nullptr, nullptr,
+            WPE_TYPE_DISPLAY,
+            static_cast<GParamFlags>(G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS));
```

Comment:
Since this has moved to WebKit, now you should use `WEBKIT_PARAM_READWRITE` so you don't need to also use `G_PARAM_STATIC_STRINGS` all over the place. I'll only leave this comment once but of course it applies to everywhere.